File: combat/src/classicthunder/combat/state/DrawingState.java
package classicthunder.combat.state;

import classicthunder.combat.hand.Discarder;
import classicthunder.combat.hand.Hand;
import classicthunder.combat.pile.DiscardPile;
import classicthunder.combat.pile.DrawPile;

public class DrawingState implements State {

    private final Hand hand;
    private final DrawPile drawPile;
    private final DiscardPile discardPile;
    private final Discarder discarder;

    private final int DRAW_TIMER = 20;
    int timer = 0;

    public DrawingState(Hand hand, DrawPile drawPile, DiscardPile discardPile, Discarder discarder) {

        this.hand = hand;
        this.drawPile = drawPile;
        this.discardPile = discardPile;
        this.discarder = discarder;
    }

    @Override
    public void Enter() {

        timer = 0;
    }

    @Override
    public EngineState Update() {

        if (timer >= DRAW_TIMER && hand.GetCardCount() < 5) {

            if (drawPile.GetCardCount() == 0) {
                drawPile.SetPile(discardPile.Empty());
            }

            hand.AddCard(drawPile.DrawTopCard());
            timer = 0;
        }

        timer++;

        hand.Update();
        discarder.Update();

        if (hand.GetCardCount() == 5 && timer == 20) {
            return EngineState.PlayerControl;
        }

        return EngineState.Drawing;
    }

    @Override
    public void Exit() {

    }
}

File: combat/src/classicthunder/combat/state/DrawingState.kt
package classicthunder.combat.state

import classicthunder.combat.hand.Discarder
import classicthunder.combat.hand.Hand
import classicthunder.combat.pile.DiscardPile
import classicthunder.combat.pile.DrawPile

internal class DrawingState(
    private val hand: Hand,
    private val drawPile: DrawPile,
    private val discardPile: DiscardPile,
    private val discarder: Discarder
) : State {
    private val DRAW_TIMER = 20
    var timer = 0
    override fun enter() {
        timer = 0
    }

    override fun update(): EngineState {
        if (timer >= DRAW_TIMER && hand.getCardCount() < 5) {
            if (drawPile.getCardCount() == 0) {
                drawPile.setPile(discardPile.empty())
            }
            hand.addCard(drawPile.drawTopCard())
            timer = 0
        }
        timer++
        hand.update()
        discarder.update()
        return if (hand.getCardCount() == 5 && timer == 20) {
            EngineState.PlayerControl
        } else EngineState.Drawing
    }

    override fun exit() {}
}